src/main/java/dk/kb/discover/webservice/KBOAuth2Handler.java

```diff
@@ -0,0 +1,498 @@
+/*
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ */
+package dk.kb.discover.webservice;
+
+import org.json.JSONObject;
+import org.json.JSONArray;
+import dk.kb.discover.config.ServiceConfig;
+import dk.kb.util.webservice.exception.InternalServiceException;
+import dk.kb.util.yaml.YAML;
+import org.apache.commons.io.IOUtils;
+import org.apache.cxf.interceptor.Fault;
+import org.json.JSONTokener;
+import org.keycloak.TokenVerifier;
+import org.keycloak.common.VerificationException;
+import org.keycloak.jose.jws.JWSInput;
+import org.keycloak.jose.jws.JWSInputException;
+import org.keycloak.representations.AccessToken;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.xml.bind.ValidationException;
+import java.io.IOException;
+import java.math.BigInteger;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.security.KeyFactory;
+import java.security.PublicKey;
+import java.security.spec.RSAPublicKeySpec;
+import java.util.Base64;
+import java.util.Collections;
+import java.util.Date;
+import java.util.HashSet;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * Caching of public keys, validation of accessTokens etc. with a focus on the parts used at the Royal Danish Library.
+ * <p>
+ * The implementation uses the <a href="https://github.com/scribejava/scribejava">Scribe library</a> for the
+ * OAuth-specific handling.
+ * <p>
+ * This class is thread safe.
+ */
+public class KBOAuth2Handler {
+    private static final Logger log = LoggerFactory.getLogger(KBOAuth2Handler.class);
+
+    public enum MODE {OFFLINE, ENABLED}
+
+    private final MODE mode;
+    private final String baseurl;
+    private final Set<String > realms;
+    private final int keysTTL;
+
+    final Map<String, PublicKey> realmKeys;
```

Comment:
Could we add a java doc on 'realms' and 'realmKeys'

Reply:
added some doc. Realms are obvious if you know how KeyCloak works.